internal/devconfig/config.go

```diff
@@ -104,19 +114,36 @@ func loadBytes(b []byte) (*Config, error) {
 }
 
 func (c *Config) LoadRecursive(lockfile *lock.File) error {
+	return c.loadRecursive(lockfile, map[string]bool{})
+}
+
+func (c *Config) loadRecursive(
+	lockfile *lock.File,
+	seen map[string]bool,
+) error {
 	included := make([]*Config, 0, len(c.Root.Include))
 
 	for _, includeRef := range c.Root.Include {
-		pluginConfig, err := plugin.LoadConfigFromInclude(includeRef, lockfile)
+		pluginConfig, err := plugin.LoadConfigFromInclude(
+			includeRef, lockfile, c.absRootPath)
 		if err != nil {
 			return errors.WithStack(err)
 		}
 
+		if seen[pluginConfig.Source.Hash()] {
+			return errors.Errorf("circular include detected: %s", includeRef)
+		}
+		seen[pluginConfig.Source.Hash()] = true
+
 		includable := &Config{
 			Root:       pluginConfig.ConfigFile,
 			pluginData: &pluginConfig.PluginOnlyData,
 		}
-		if err := includable.LoadRecursive(lockfile); err != nil {
+		if hasDir, ok := pluginConfig.Source.(*plugin.LocalPlugin); ok {
+			includable.absRootPath = filepath.Dir(hasDir.Path())
+		}
```

Comment:
the name `hasDir` indicates to me that its type is boolean. Could we call it `localPlugin` instead?

Also consider encapsulating lines 138-144 into a function such as: `initIncludableFromPlugin(pluginConfig)`